File: src/LinkPulse/LinkPulse.razor.cs
using System.Text.Json;
using LinkPulse.Abstractions;
using LinkPulse.Measurement;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace LinkPulse;

/// <summary>
/// The client-facing LinkPulse component (v1 spec, &#167;8). It measures the real-time quality of
/// the browser's connection &#8212; RTT, jitter, and packet loss over a rolling window &#8212;
/// derives a stable quality rating, reports snapshots to the server probe endpoint, and renders the
/// result as a compact badge, an expanded panel, or nothing at all.
/// </summary>
/// <remarks>
/// <para>
/// The same code path runs in both Blazor Auto phases (Server and WebAssembly), so the numbers stay
/// comparable across the transition (&#167;2). All timing comes from a collocated JS module via
/// <c>performance.now()</c>; this component owns the <see cref="MeasurementEngine"/>, the quality
/// classification, the hysteresis filter, the snapshot cadence, and the rendering. It is
/// single-threaded by construction &#8212; every JS callback arrives on the renderer's
/// synchronization context &#8212; so no locking is needed.
/// </para>
/// <para>
/// JS interop only starts once the component is interactive (after the first render), so static
/// prerender shows the initial "connecting" markup and live measurement begins on hydration.
/// </para>
/// </remarks>
public sealed partial class LinkPulse : IAsyncDisposable
{
    private const string ModulePath = "./_content/LinkPulse/LinkPulse.razor.js";
    private const double SparklineWidth = 120;
    private const double SparklineHeight = 32;

    private MeasurementEngine _engine = null!;
    private QualityStabilizer _stabilizer = null!;
    private ClientPhase _phase;

    private DotNetObjectReference<LinkPulse>? _selfRef;
    private IJSObjectReference? _module;
    private IJSObjectReference? _probe;
    private bool _disposed;

    private Guid _clientId;
    private Guid _sessionId;

    private bool _connected;
    private bool _expanded;
    private MetricSnapshot? _snapshot;
    private double _nowMs;
    private double _lastReportAtMs;
    private double? _connectedSinceMs;
    private double? _downSinceMs;

    [Inject]
    private IJSRuntime JS { get; set; } = null!;

    /// <summary>Visible-tab ping cadence, in milliseconds (&#167;7). Bounded below server-side.</summary>
    [Parameter]
    public int PingIntervalMs { get; set; } = 1000;

    /// <summary>Hidden-tab ping cadence, in milliseconds, used while the tab is backgrounded (&#167;3.4).</summary>
    [Parameter]
    public int HiddenTabPingIntervalMs { get; set; } = 5000;

    /// <summary>Rolling sample count over which all derived metrics are computed (&#167;3.4).</summary>
    [Parameter]
    public int WindowSize { get; set; } = 30;

    /// <summary>A ping with no echo within this many milliseconds is counted as lost (&#167;3.2).</summary>
    [Parameter]
    public int PingTimeoutMs { get; set; } = 5000;

    /// <summary>RFC 3550 jitter gain denominator <c>G</c> (&#167;3.3).</summary>
    [Parameter]
    public double JitterSmoothingFactor { get; set; } = 16;

    /// <summary>Snapshot reporting cadence, in milliseconds (&#167;3.5).</summary>
    [Parameter]
    public int SnapshotIntervalMs { get; set; } = 5000;

    /// <summary>Consecutive snapshots a new rating must hold before the display changes (&#167;6).</summary>
    [Parameter]
    public int HysteresisSamples { get; set; } = 3;

    /// <summary>How the component renders: <see cref="DisplayMode.Badge"/>, <see cref="DisplayMode.Panel"/>, or <see cref="DisplayMode.Hidden"/>.</summary>
    [Parameter]
    public DisplayMode Display { get; set; } = DisplayMode.Badge;

    /// <summary>Quality-classification tier boundaries (&#167;6); defaults to <see cref="QualityThresholds.Default"/>.</summary>
    [Parameter]
    public QualityThresholds Thresholds { get; set; } = QualityThresholds.Default;

    /// <summary>The WebSocket probe path the client connects to; resolved against the app origin (&#167;12).</summary>
    [Parameter]
    public string ProbePath { get; set; } = "/connection-probe";

    /// <summary>The rating currently displayed: the stabilized measured rating, or <see cref="QualityRating.Disconnected"/> while the probe is down.</summary>
    private QualityRating DisplayedRating => _connected ? _stabilizer.Current : QualityRating.Disconnected;

    /// <inheritdoc />
    protected override void OnInitialized()
    {
        _phase = OperatingSystem.IsBrowser() ? ClientPhase.Wasm : ClientPhase.Server;
        _engine = new MeasurementEngine(WindowSize, PingTimeoutMs, JitterSmoothingFactor);
        _stabilizer = new QualityStabilizer(HysteresisSamples);
    }

    /// <inheritdoc />
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (!firstRender || _disposed)
        {
            return;
        }

        _selfRef = DotNetObjectReference.Create(this);
        try
        {
            _module = await JS.InvokeAsync<IJSObjectReference>("import", ModulePath);
            if (_disposed)
            {
                return; // disposed mid-import
            }

            _probe = await _module.InvokeAsync<IJSObjectReference>(
                "createProbe", _selfRef, ProbePath, PingIntervalMs, HiddenTabPingIntervalMs);
        }
        catch (JSException)
        {
            // The module failed to load or initialise (network error, blocked dynamic import, …).
            // A telemetry widget must not take down its host page, so stay in the disconnected
            // state and simply do not measure.
        }
    }

    /// <summary>JS callback: the probe socket (re)connected. Adopts the identity and re-arms measurement.</summary>
    /// <param name="clientId">The stable per-browser identifier from <c>localStorage</c>.</param>
    /// <param name="sessionId">The fresh per-connection identifier.</param>
    /// <param name="nowMs">The client-clock time of the connect, in milliseconds.</param>
    [JSInvokable]
    public void OnConnected(string clientId, string sessionId, double nowMs)
    {
        // These ids are always freshly-minted UUID strings from the JS module, so the parse cannot
        // realistically fail; the guard merely keeps a malformed value from throwing, leaving an
        // empty Guid rather than crashing the connect.
        _ = Guid.TryParse(clientId, out _clientId);
        _ = Guid.TryParse(sessionId, out _sessionId);

        // A new connection is a new measurement session (fresh SessionId): start the engine clean so
        // the dropped connection's stale window and still-outstanding pings cannot bleed into — and
        // be counted as loss against — the new one.
        _engine = new MeasurementEngine(WindowSize, PingTimeoutMs, JitterSmoothingFactor);

        _connected = true;
        _nowMs = nowMs;
        _connectedSinceMs = nowMs;
        _downSinceMs = null;
        _lastReportAtMs = nowMs;
        _stabilizer.Reset(); // first measured rating after (re)connect is adopted at once (§6 seam)
        StateHasChanged();
    }

    /// <summary>JS callback: a ping was sent. Hands the send-stamp to the engine.</summary>
    /// <param name="seq">The ping's monotonic sequence number.</param>
    /// <param name="sentAtMs">The client-clock send time, in milliseconds.</param>
    [JSInvokable]
    public void OnPing(long seq, double sentAtMs) => _engine.RecordPing(seq, sentAtMs);

    /// <summary>JS callback: an echo arrived. Pairs it to its ping by sequence number.</summary>
    /// <param name="seq">The echoed sequence number.</param>
    /// <param name="receivedAtMs">The client-clock receive time, in milliseconds.</param>
    [JSInvokable]
    public void OnEcho(long seq, double receivedAtMs) => _engine.RecordEcho(seq, receivedAtMs);

    /// <summary>
    /// JS callback fired on the adaptive cadence: ages out timed-out pings, reports a snapshot when
    /// one is due, and refreshes the live counters.
    /// </summary>
    /// <param name="nowMs">The current client-clock time, in milliseconds.</param>
    [JSInvokable]
    public async Task OnTick(double nowMs)
    {
        _nowMs = nowMs;
        _engine.ExpireOutstanding(nowMs);

        if (!_connected)
        {
            _downSinceMs ??= nowMs; // count up from first tick even before the first connect
        }
        else if (nowMs - _lastReportAtMs >= SnapshotIntervalMs)
        {
            await ReportSnapshotAsync(nowMs);
        }

        StateHasChanged();
    }

    /// <summary>JS callback: the probe socket dropped. Freezes the last-known metrics and starts the reconnect count-up.</summary>
    /// <param name="nowMs">The client-clock time of the drop, in milliseconds.</param>
    [JSInvokable]
    public void OnDisconnected(double nowMs)
    {
        _connected = false;
        _nowMs = nowMs;
        _downSinceMs = nowMs;
        _connectedSinceMs = null;
        StateHasChanged();
    }

    private async Task ReportSnapshotAsync(double nowMs)
    {
        var snapshot = _engine.CreateSnapshot(_phase);
        _snapshot = snapshot;
        _stabilizer.Push(QualityCalculator.Classify(snapshot, Thresholds));
        _lastReportAtMs = nowMs;

        if (_probe is null)
        {
            return;
        }

        ProbeFrame frame = SnapshotFrame.FromSnapshot(_clientId, _sessionId, snapshot);
        var json = JsonSerializer.Serialize(frame, LinkPulseJsonContext.Default.ProbeFrame);
        try
        {
            await _probe.InvokeVoidAsync("send", json);
        }
        catch (JSException)
        {
            // Best-effort telemetry. A torn-down circuit (JSDisconnectedException) or a transient
            // interop fault should drop this one snapshot — never propagate out of this JSInvokable
            // callback and fault the circuit, which would end the user's whole session.
        }
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        _disposed = true;
        try
        {
            if (_probe is not null)
            {
                await _probe.InvokeVoidAsync("dispose");
                await _probe.DisposeAsync();
            }

            if (_module is not null)
            {
                await _module.DisposeAsync();
            }
        }
        catch (Exception ex) when (ex is JSDisconnectedException or OperationCanceledException)
        {
            // The circuit/renderer is already being torn down; the JS side is disposed with it.
        }

        _selfRef?.Dispose();
    }
}

File: src/LinkPulse/LinkPulse.razor.js
// LinkPulse.razor.js — collocated RCL module for the <LinkPulse /> component.
//
// This module is deliberately "dumb": it owns only the browser-side concerns that cannot live in
// .NET — the raw WebSocket probe channel, high-resolution timing (performance.now), the Page
// Visibility adaptive cadence, reconnect backoff, and the localStorage client identity. Every
// timestamp it reads is handed to .NET verbatim; all metric math (RTT, jitter, loss, quality)
// runs in the C# MeasurementEngine, which is exactly why that engine takes its time from outside.
//
// One probe instance per <LinkPulse /> component; createProbe returns a handle the component holds
// as an IJSObjectReference and calls send()/dispose() on.

const CLIENT_ID_KEY = "linkpulse.clientId";

// A UUID that works everywhere. crypto.randomUUID() is the happy path but is only defined in secure
// contexts (HTTPS / localhost), so plain-http intranet hosting would throw; fall back to an RFC 4122
// v4 built from crypto.getRandomValues (available in insecure contexts too), then to Math.random as
// a last resort. This never throws, so identity is always available.
function newId() {
    try {
        if (typeof crypto !== "undefined" && crypto.randomUUID) {
            return crypto.randomUUID();
        }
    } catch {
        // secure-context-only; fall through
    }

    const bytes = new Uint8Array(16);
    if (typeof crypto !== "undefined" && crypto.getRandomValues) {
        crypto.getRandomValues(bytes);
    } else {
        for (let i = 0; i < 16; i++) {
            bytes[i] = Math.floor(Math.random() * 256);
        }
    }
    bytes[6] = (bytes[6] & 0x0f) | 0x40; // version 4
    bytes[8] = (bytes[8] & 0x3f) | 0x80; // variant 10
    const hex = [...bytes].map((b) => b.toString(16).padStart(2, "0"));
    return `${hex.slice(0, 4).join("")}-${hex.slice(4, 6).join("")}-${hex.slice(6, 8).join("")}-${hex.slice(8, 10).join("")}-${hex.slice(10, 16).join("")}`;
}

// Stable per-browser id (spec §4): minted once, persisted, survives reloads and the Server→WASM
// transition. Falls back to an ephemeral id when storage is unavailable (private mode), so
// measurement still works — only cross-reload grouping is lost.
function getClientId() {
    try {
        let id = localStorage.getItem(CLIENT_ID_KEY);
        if (!id) {
            id = newId();
            localStorage.setItem(CLIENT_ID_KEY, id);
        }
        return id;
    } catch {
        return newId();
    }
}

// Resolves the probe path against the document base and forces the ws/wss scheme to match the page,
// so a relative "/connection-probe" reaches the app's own origin over the right transport.
function probeUrl(probePath) {
    const url = new URL(probePath, document.baseURI);
    url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
    return url.toString();
}

export function createProbe(dotNet, probePath, pingIntervalMs, hiddenTabPingIntervalMs) {
    const clientId = getClientId();
    const initialReconnectMs = 1000;
    const maxReconnectMs = 30000;

    let socket = null;
    let sessionId = null;
    let seq = 0;
    let connected = false;
    let disposed = false;
    let reconnectAttempt = 0;
    let tickTimer = null;
    let reconnectTimer = null;

    // Fire-and-forget interop: the .NET object reference may already be torn down when a callback
    // lands (component disposed mid-flight), so swallow the resulting rejection.
    const notify = (method, ...args) => {
        if (disposed) {
            return;
        }
        dotNet.invokeMethodAsync(method, ...args).catch(() => { });
    };

    const currentInterval = () =>
        document.visibilityState === "hidden" ? hiddenTabPingIntervalMs : pingIntervalMs;

    // A single self-rescheduling loop drives everything. Its period is the adaptive ping cadence
    // (1 Hz visible / 5 s hidden, §3.4); .NET decides snapshot timing itself from the timestamps,
    // so the loop needs only to (a) send a ping when connected and (b) hand .NET the current clock.
    function tick() {
        if (disposed) {
            return;
        }

        const now = performance.now();
        if (connected && socket && socket.readyState === WebSocket.OPEN) {
            seq += 1;
            try {
                socket.send(JSON.stringify({ type: "ping", seq, payload: now.toString() }));
                notify("OnPing", seq, now);
            } catch {
                // Socket dropped between the readyState check and send; onclose will handle it.
            }
        }

        notify("OnTick", now);
        tickTimer = setTimeout(tick, currentInterval());
    }

    function connect() {
        if (disposed) {
            return;
        }

        let ws;
        try {
            ws = new WebSocket(probeUrl(probePath));
        } catch {
            scheduleReconnect();
            return;
        }

        socket = ws;

        ws.onopen = () => {
            if (disposed) {
                try { ws.close(); } catch { /* ignore */ }
                return;
            }
            connected = true;
            reconnectAttempt = 0;
            sessionId = newId(); // fresh per connection (spec §4)
            notify("OnConnected", clientId, sessionId, performance.now());
        };

        ws.onmessage = (event) => {
            const t1 = performance.now();
            let frame;
            try {
                frame = JSON.parse(event.data);
            } catch {
                return; // not our frame
            }
            // Echo of our ping, matched by seq in .NET (out-of-order safe). payload is ignored here:
            // .NET already holds t0 keyed by seq, so only the receive stamp is needed.
            if (frame && frame.type === "ping" && typeof frame.seq === "number") {
                notify("OnEcho", frame.seq, t1);
            }
        };

        ws.onerror = () => {
            try { ws.close(); } catch { /* onclose follows */ }
        };

        ws.onclose = () => {
            if (socket !== ws) {
                return; // superseded by a newer socket
            }
            const wasConnected = connected;
            connected = false;
            socket = null;
            if (wasConnected) {
                notify("OnDisconnected", performance.now());
            }
            scheduleReconnect();
        };
    }

    // Exponential backoff with full jitter (spec §5.2): 1→2→4…s, capped at 30 s, retried forever.
    // Math.min absorbs the eventual 2**n → Infinity, so the cap holds without overflow.
    function scheduleReconnect() {
        if (disposed) {
            return;
        }
        const ceiling = Math.min(maxReconnectMs, initialReconnectMs * 2 ** reconnectAttempt);
        reconnectAttempt += 1;
        reconnectTimer = setTimeout(connect, Math.random() * ceiling);
    }

    // Becoming visible should resume the fast cadence at once rather than waiting out a 5 s hidden
    // tick, so cancel the pending tick and fire immediately.
    function onVisibility() {
        if (!disposed && document.visibilityState === "visible") {
            if (tickTimer) {
                clearTimeout(tickTimer);
            }
            tick();
        }
    }

    document.addEventListener("visibilitychange", onVisibility);
    connect();
    tick();

    return {
        // .NET pushes a fully-serialized snapshot frame; we just put it on the wire when open.
        send(json) {
            if (socket && socket.readyState === WebSocket.OPEN) {
                try { socket.send(json); } catch { /* drop; next snapshot will follow */ }
            }
        },
        dispose() {
            disposed = true;
            document.removeEventListener("visibilitychange", onVisibility);
            if (tickTimer) {
                clearTimeout(tickTimer);
            }
            if (reconnectTimer) {
                clearTimeout(reconnectTimer);
            }
            if (socket) {
                try { socket.close(); } catch { /* ignore */ }
            }
            socket = null;
        },
    };
}
